lib/spack/spack/directives.py

```diff
@@ -640,14 +643,30 @@ def _raise_default_not_set(pkg):
     description = str(description).strip()
 
     def _execute_variant(pkg):
+        when_spec = make_when_spec(when)
+        when_specs = [when_spec]
+
         if not re.match(spack.spec.identifier_re, name):
             directive = 'variant'
             msg = "Invalid variant name in {0}: '{1}'"
             raise DirectiveError(directive, msg.format(pkg.name, name))
 
-        pkg.variants[name] = spack.variant.Variant(
+        variant_desc = spack.variant.Variant(name, default, description,
+                                             values, multi, validator)
+
+        if name in pkg.variants:
+            orig_variant, orig_when = pkg.variants[name]
+            if variant_desc != orig_variant:
+                msg = ("multiply defined variant %s " % name +
+                       "has incompatible declarations. Only when-clauses " +
+                       "may differ between variant declarations.")
+                raise DirectiveError(msg)
```

Comment:
I added the following to `cp2k`:
```python
variant('foo', default=True, description='', when='+mpi')
variant('foo', default=False, description='', when='+openmp')
```
but I couldn't trigger this error.

Reply:
Can we remove this error check completely? With the comparison bug solved, this will not permit overriding a variant from a base package:
```console
$ spack solve cp2k
==> Error: cannot bootstrap the "clingo" Python module from spec "clingo-bootstrap@spack+python %gcc target=x86_64" due to the following failures:
    'github-actions' raised DirectiveError: multiply defined variant build_type has incompatible declarations. Only when-clauses may differ between variant declarations.
    'spack-install' raised DirectiveError: multiply defined variant build_type has incompatible declarations. Only when-clauses may differ between variant declarations.
    Please run `spack -d spec zlib` for more verbose error messages
```

Reply:
Solved in ca5eb36